src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -104,21 +104,32 @@ public Task<IEnumerable<string>> PrerenderComponentAsync(Type componentType, Par
             });
         }
 
-        public async Task InitializeAsync(CancellationToken cancellationToken)
+        public void Initialize(CancellationToken cancellationToken)
         {
-            await Renderer.InvokeAsync(async () =>
+            // This Initialize method is fire-and-forget as far as the caller is concerned, because
+            // if it was to await completion, it would be blocking the SignalR message loop. This could
+            // lead to deadlock, e.g., if the init process itself waited for an incoming SignalR message
+            // such as the result of a JSInterop call.
+            Renderer.InvokeAsync(async () =>
             {
-                SetCurrentCircuitHost(this);
-
-                for (var i = 0; i < Descriptors.Count; i++)
+                try
                 {
-                    var (componentType, domElementSelector) = Descriptors[i];
-                    await Renderer.AddComponentAsync(componentType, domElementSelector);
-                }
+                    SetCurrentCircuitHost(this);
 
-                await OnCircuitOpenedAsync(cancellationToken);
+                    for (var i = 0; i < Descriptors.Count; i++)
+                    {
+                        var (componentType, domElementSelector) = Descriptors[i];
+                        await Renderer.AddComponentAsync(componentType, domElementSelector);
```

Comment:
Do we still have to work for quiescence here until we declare the circuit opened?

Reply:
No, if we did that, then JSInterop calls made in the meantime (e.g., during InitAsync) would throw because they are only allowed once the circuit is flagged as opened.

We need the circuit to be regarded as opened synchronously. Since dispatcher calls are serialized, and since (non-quiescent) rendering is synchronous, we already know that we'll only start to process incoming JSInterop calls *after* the first synchronous render, which is the right time to start processing them.